plasmapy/simulation/particle_tracker.py

```diff
@@ -0,0 +1,945 @@
+        super().__init__(**kwargs)
+        self.t_all = []
+
+        self.save_interval = interval.to(u.s).value
+        self.time_of_last_save = 0
+
+    @property
+    def require_synchronized_dt(self):
+        """Save output only makes sense for synchronized time steps."""
+        return True
+
+    @property
+    def save_now(self):
+        """Save at every interval given in instantiation."""
+
+        return self.tracker.time - self.time_of_last_save >= self.save_interval
+
+    def save(self):
+        """Save the current state of the simulation.
+        Sets the time of last save attribute and log the timestamp.
+        """
+        super().save()
+
+        self.time_of_last_save = self.tracker.time
+        self.t_all.append(self.tracker.time)
+
+
+class ParticleTracker:
+    r"""A particle tracker for particles in electric and magnetic fields without inter-particle interactions.
+
+    Particles are instantiated and pushed through a grid of provided E and B fields
+    using the boris push algorithm. These fields are specified as part of a grid
+    which are then interpolated to determine the local field acting on each particle.
+
+    The time step used in the push routine can be specified, or an adaptive time step
+    will be determined based off the gyroradius of the particle. Some save routines involve
+    time stamping the location and velocities of particles at fixed intervals. In order for this
+    data to make sense, it is required that all particles follow the same time step. This is
+    referred to as a synchronized time step. If no time step is specified and the provided
+    save routine does not require a synchronized time step, then an adaptive time step is calculated
+    independently for each particle.
+
+    The time step used can be specified, otherwise an
+    adaptive time step is calculated based off grid resolution and the particle's gyroradius.
+
+    Termination conditions for the simulation are provided as an instance of the `~plasmapy.simulation.particle_tracker.AbstractTerminationCondition` class as arguments to the ``run`` method.
+    The results of a simulation can be exported by specifying an instance of the `~plasmapy.simulation.particle_tracker.AbstractSaveRoutine` class to the ``run`` method.
+
+    Parameters
+    ----------
+    grids : An instance of `~plasmapy.plasma.grids.AbstractGrid`
+        A Grid object or list of grid objects containing the required
+        quantities.
+
+    termination_condition : `~plasmapy.simulation.particle_tracker.AbstractTerminationCondition`
+        A class responsible for determining when the simulation has reached
+        a suitable termination condition. The class is passed an instance
+        of `~plasmapy.simulation.particle_tracker.ParticleTracker`. See `~plasmapy.simulation.particle_tracker.AbstractTerminationCondition`
+        for more details.
+
+    save_routine : `~plasmapy.simulation.particle_tracker.AbstractSaveRoutine`
```

Comment:
For each of these, add a sentence saying what the default value is, and what does.